src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java

```diff
@@ -147,71 +152,103 @@ public MultiClusterPooledConnectionProvider(MultiClusterClientConfig multiCluste
+                healthStatusManager.add(config.getHostAndPort(), hcs);
             }
         }
 
+        // selecting activeCluster with configuration values.
+        // all health status would be HEALTHY at this point
+        activeCluster = findWeightedHealthyCluster().getValue();
+
+        for (Endpoint endpoint : multiClusterMap.keySet()) {
+            healthStatusManager.registerListener(endpoint, this::handleStatusChange);
+        }
         /// --- ///
 
         this.fallbackExceptionList = multiClusterClientConfig.getFallbackExceptionList();
     }
 
-    /**
-     * Increments the actively used cluster/database endpoint (connection pool) amongst the pre-configured list which
-     * were provided at startup via the MultiClusterClientConfig. All traffic will be routed according to this index.
-     * Only indexes within the pre-configured range (static) are supported otherwise an exception will be thrown. In the
-     * event that the next prioritized connection has a forced open state, the method will recursively increment the
-     * index in order to avoid a failed command.
-     */
-    public int incrementActiveMultiClusterIndex() {
+    private void handleStatusChange(HealthStatusChangeEvent eventArgs) {
 
-        // Field-level synchronization is used to avoid the edge case in which
-        // setActiveMultiClusterIndex(int multiClusterIndex) is called at the same time
-        activeClusterIndexLock.lock();
+        Endpoint endpoint = eventArgs.getEndpoint();
+        HealthStatus newStatus = eventArgs.getNewStatus();
+        log.info("Health status changed for {} from {} to {}", endpoint, eventArgs.getOldStatus(), newStatus);
 
-        try {
-            String originalClusterName = getClusterCircuitBreaker().getName();
+        Cluster clusterWithHealthChange = multiClusterMap.get(endpoint);
+
+        if (clusterWithHealthChange == null) return;
+        if (clusterWithHealthChange.isForcedUnhealthy()) return;
 
-            // Only increment if it can pass this validation otherwise we will need to check for NULL in the data path
-            if (activeMultiClusterIndex + 1 > multiClusterMap.size()) {
+        clusterWithHealthChange.setHealthStatus(newStatus);
 
-                lastClusterCircuitBreakerForcedOpen = true;
+        if (newStatus.isHealthy()) {
+            if (clusterWithHealthChange.isFailbackEnabled() && activeCluster != clusterWithHealthChange) {
+                // lets check if weighted switching is possible
+                Map.Entry<Endpoint, Cluster> failbackCluster = findWeightedFailbackCluster();
+                if (failbackCluster == clusterWithHealthChange
+                    && clusterWithHealthChange.getWeight() > activeCluster.getWeight()) {
+                    setActiveCluster(clusterWithHealthChange, false);
+                }
+            }
+        } else if (clusterWithHealthChange == activeCluster) {
+            iterateActiveCluster();
+        }
```

Comment:
@a-TODO-rov, is this what you suggest? if not, could you provide code snippet on your suggestion? 

```suggestion
        if (clusterWithHealthChange == activeCluster && !newStatus.isHealthy()) {
            iterateActiveCluster();
            return;
        }
        if (newStatus.isHealthy() && clusterWithHealthChange.isFailbackEnabled() && activeCluster != clusterWithHealthChange) {
                // lets check if weighted switching is possible
                Map.Entry<Endpoint, Cluster> failbackCluster = findWeightedFailbackCluster();
                if (failbackCluster == clusterWithHealthChange
                    && clusterWithHealthChange.getWeight() > activeCluster.getWeight()) {
                    setActiveCluster(clusterWithHealthChange, false);
                }
            }
        } 
```